### PR TITLE
Improve usability of -l flag

### PR DESCRIPTION
**Pull request by RAJAGOPALAN-GANGADHARAN**
Resolves [SR-14122](https://bugs.swift.org/browse/SR-14122)

Complements PR from [bundled driver at swift](https://github.com/apple/swift/pull/38750)

### CODE REVIEW

**Comment by RAJAGOPALAN-GANGADHARAN** [comment]
Requesting review : @varungandhi-apple

**Comment by RAJAGOPALAN-GANGADHARAN** [comment]
Should I use swift-format to format these code? I'm guessing clang-format is not suitable for this

File: Sources/SwiftDriver/Jobs/CommandLineArguments.swift [hunk 1]
```diff
@@ -149,6 +149,24 @@ extension Array where Element == Job.ArgTemplate {
...
+        var isExcluded = false
+        for exclude in excludeList {
+          if optGroup.option == exclude {
+            isExcluded = true
+            break
+          }
+        }
+        if !isExcluded {
+          try append(optGroup)
+        }
```
Comment (varungandhi-apple) [inline Sources/SwiftDriver/Jobs/CommandLineArguments.swift:165]: You could simplify this logic to be something like:

```
if !excludeList.contains { $0 == optGroup.option } {
  try append(optGroup) 
}
```

**Comment by varungandhi-apple** [comment]
Formatting-wise, we're not using swift-format, but please follow the style of the surrounding code. For example, there should be a space before an opening brace.

File: Sources/SwiftDriver/Jobs/DarwinToolchain+LinkerSupport.swift [hunk 2]
```diff
@@ -422,8 +422,14 @@ extension DarwinToolchain {
 
     // These custom arguments should be right before the object file at the
     // end.
-    try commandLine.append(
-      contentsOf: parsedOptions.arguments(in: .linkerOption)
+    try commandLine.appendAllExcept(
+      includeList:[.linkerOption],
+      excludeList:[.l],
```
Comment (varungandhi-apple) [inline Sources/SwiftDriver/Jobs/DarwinToolchain+LinkerSupport.swift:427]: Nit: There should be a space after the `:` (also applicable elsewhere).

File: Sources/SwiftDriver/Jobs/Toolchain+LinkerSupport.swift [hunk 3]
```diff
@@ -86,6 +86,15 @@ extension Toolchain {
     ).appending(component: runtimeName)
     return try fileSystem.exists(path)
   }
+
+  func addLinkedLibArgs(
+    to commandLine: inout [Job.ArgTemplate],
+    parsedOptions: inout ParsedOptions
+  ) {
+    for match in parsedOptions.arguments(for: .l) {
+      commandLine.appendFlag(match.option.spelling+match.argument.asSingle)
```
Comment (varungandhi-apple) [inline Sources/SwiftDriver/Jobs/Toolchain+LinkerSupport.swift:95]: Nit: Please use spaces on both sides of the `+`.

File: Tests/SwiftDriverTests/SwiftDriverTests.swift [hunk 4]
```diff
@@ -2302,6 +2318,13 @@ final class SwiftDriverTests: XCTestCase {
       return false
     }
 
+    func isSquashedArgContains(arg: Job.ArgTemplate, checkOpt: Job.ArgTemplate) -> Bool {
+      if case let .squashedArgumentList(option: _, args: args) = arg {
+        return args.contains(checkOpt)
+      }
+      return false
+    }
```
Comment (varungandhi-apple) [inline Tests/SwiftDriverTests/SwiftDriverTests.swift:2326]: Since this function is only being used in one place 2 times, I think you could inline it and do checks like:

```
guard case let .squashedArgumentList(option: _, args: args) = cmd[0] {
  // XCTFail
}
XCTAssertTrue(args.contains(.flag(...)))
XCTAssertTrue(args.contains(.flag(...)))
```

File: Tests/SwiftDriverTests/SwiftDriverTests.swift [hunk 5]
```diff
@@ -2346,6 +2369,20 @@ final class SwiftDriverTests: XCTestCase {
...
+    do{
+      // Linked library arguments with space
+      var driver = try Driver(args: ["swift", "-repl", "-l", "somelib", "-lotherlib"])
+      let plannedJobs = try driver.planBuild()
+      XCTAssertEqual(plannedJobs.count, 1)
+      let replJob = plannedJobs.first!
+      XCTAssertTrue(replJob.tool.name.contains("lldb"))
+      XCTAssertTrue(replJob.requiresInPlaceExecution)
+      let cmd = replJob.commandLine
+      XCTAssert(cmd.contains(where: { isExpectedLLDBREPLFlag($0) }))
```
Comment (varungandhi-apple) [inline Tests/SwiftDriverTests/SwiftDriverTests.swift:2382]: I think the checks here for `requiresInPlaceExecution` and `isExpectedLLDBREPLFlag` are not needed, since those are tested elsewhere, and the main thing this bit is trying to check is whether the `-l` arguments are passed correctly.

File: Tests/SwiftDriverTests/SwiftDriverTests.swift [hunk 6]
```diff
@@ -1016,6 +1016,22 @@ final class SwiftDriverTests: XCTestCase {
...
+      XCTAssertFalse(plannedJobs.contains { $0.kind == .autolinkExtract })
+
+      let linkJob = plannedJobs[2]
+      XCTAssertEqual(linkJob.kind, .link)
+
+      let cmd = linkJob.commandLine
+      XCTAssertTrue(cmd.contains(.flag("-lsomelib")))
+      XCTAssertTrue(cmd.contains(.flag("-lotherlib")))
+    }
+
```
Comment (varungandhi-apple) [inline Tests/SwiftDriverTests/SwiftDriverTests.swift:1034]: Couple of things here:

1. Right now, this `do` block is at the start of the test, but arguably there are tests for common cases below it. I would try to put this near the end, since it's testing edge-case behavior.
2. Given that multiple other tests are checking for `$0.kind == .autolinkExtract` and `XCTAssertEqual(linkJob.kind, .link)`, I'm not sure if that's really needed here.

**Comment by varungandhi-apple** [comment]
Largely LGTM. Most of the comments are kinda' nitpicks (sorry!) about placement/code simplication. Thank you for your patience and submitting this mirroring PR.

**Comment by RAJAGOPALAN-GANGADHARAN** [comment]
Thank you for reviewing the code and for your patient feedback. Maybe my next PR would be little bit better I guess :)

**Comment by RAJAGOPALAN-GANGADHARAN** [comment]
Could someone trigger swift-ci for this please?

**Comment by artemcm** [comment]
@swift-ci please test

**Comment by artemcm** [comment]
@swift-ci please test
